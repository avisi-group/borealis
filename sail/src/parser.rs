//! Sail source file parsing

use crate::{
    ast::{type_check::Env, Ast},
    error::Error,
    RT,
};

/// Parses supplied Sail files and returns the AST
///
/// From Sail internal docs:
///
/// This function parses all the files passed to Sail, and then concatenates
/// their ASTs. The pre-processor is then run, which evaluates `$directive`
/// statements in Sail, such as
///
/// ```sail
/// $include <prelude.sail>
/// ```
///
/// Unlike the C pre-processor the Sail pre-processor operates over actual
/// Sail ASTs rather than strings. This can recursively include other
/// files into the AST, as well as add/remove parts of the AST with
/// `$ifdef` etc. Directives that are used are preserved in the AST, so
/// they also function as a useful way to pass auxiliary information to
/// the various Sail backends.
///
/// The initial check mentioned above is then run to desugar the AST, and
/// then the type-checker is run which produces a fully type-checked
/// AST. Type annotations are attached to every node (for which an
/// annotation makes sense) using the aux constructors.
///
/// After type-checking the Sail scattered definitions are de-scattered
/// into single functions.
pub fn load_files(files: Vec<String>) -> Result<(String, Ast, Env), Error> {
    Ok(RT.lock().load_files(files)?)
}

#[cfg(test)]
mod tests {
    use {
        crate::parser::load_files,
        std::{env, path::PathBuf},
    };

    #[test]
    fn load_files_empty() {
        insta::assert_debug_snapshot!(load_files(vec![]).unwrap());
    }

    #[test]
    fn load_files_prelude() {
        let path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap())
            .join("examples/prelude.sail")
            .to_string_lossy()
            .to_string();

        insta::assert_debug_snapshot!(load_files(vec![path]).unwrap());
    }
}
