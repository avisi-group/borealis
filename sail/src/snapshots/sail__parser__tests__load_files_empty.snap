---
source: sail/src/parser.rs
expression: "load_files(vec![]).unwrap()"
---
(
    "out.sail",
    Ast {
        defs: [],
        comments: [],
    },
    Env {
        top_val_specs: Root(
            Root(
                0x0000000130540320,
            ),
        ),
        defined_val_specs: Root(
            Root(
                0x0000000130540328,
            ),
        ),
        locals: Root(
            Root(
                0x0000000130540330,
            ),
        ),
        top_letbinds: Root(
            Root(
                0x0000000130540338,
            ),
        ),
        union_ids: Root(
            Root(
                0x0000000130540340,
            ),
        ),
        registers: Root(
            Root(
                0x0000000130540348,
            ),
        ),
        variants: Root(
            Root(
                0x0000000130540350,
            ),
        ),
        scattered_variant_envs: Root(
            Root(
                0x0000000130540358,
            ),
        ),
        mappings: Root(
            Root(
                0x0000000130540360,
            ),
        ),
        typ_vars: Root(
            Root(
                0x0000000130540368,
            ),
        ),
        shadow_vars: Root(
            Root(
                0x0000000130540370,
            ),
        ),
        typ_synonyms: Root(
            Root(
                0x0000000130540378,
            ),
        ),
        overloads: Root(
            Root(
                0x0000000130540380,
            ),
        ),
        enums: Root(
            Root(
                0x0000000130540388,
            ),
        ),
        records: Root(
            Root(
                0x0000000130540390,
            ),
        ),
        accessors: Root(
            Root(
                0x0000000130540398,
            ),
        ),
        externs: Root(
            Root(
                0x00000001305403a0,
            ),
        ),
        casts: [],
        allow_casts: true,
        allow_bindings: true,
        constraints: [],
        default_order: None,
        ret_typ: None,
        poly_undefineds: false,
        prove: Some(
            Root(
                Root(
                    0x00000001305403b0,
                ),
            ),
        ),
        allow_unknowns: false,
        bitfields: Root(
            Root(
                0x00000001305403a8,
            ),
        ),
    },
)
